### PR TITLE
Re-process npm component metadata when it expires from the metadata cache CLM-24065

### PR DESCRIPTION
**Pull request by vladt**
https://issues.sonatype.org/browse/CLM-24065
https://jenkins.ci.sonatype.dev/job/insight/job/firewall-for-artifactory/job/feature-snapshots/job/CLM-24065_ReprocessNpmMetadata/

### CODE REVIEW

File: firewall/src/main/java/com/sonatype/iq/artifactory/DefaultNexusFirewallForArtifactory.groovy [hunk 1]
```diff
@@ -88,8 +88,12 @@ class DefaultNexusFirewallForArtifactory
 
   final long verifyInitWaitInMillis
 
-  DefaultNexusFirewallForArtifactory(final Logger log, final File pluginDirectory, final Repositories repositories,
-                                     final Searches searches, final String artifactoryVersion)
+  DefaultNexusFirewallForArtifactory(
+      final Logger log,
+      final File pluginDirectory,
+      final Repositories artifactoryRepositories,
```
Comment (vladt) [inline firewall/src/main/java/com/sonatype/iq/artifactory/DefaultNexusFirewallForArtifactory.groovy:94]: Only renamed the `repositories` param to `artifactoryRepositories` to disambiguate it.

File: firewall/src/main/java/com/sonatype/iq/artifactory/NpmComponentMetadataProcessor.java [hunk 2]
```diff
@@ -58,50 +58,43 @@ public NpmComponentMetadataProcessor(IqConnectionManager iqConnectionManager) {
...
-        RepositoryComponentEvaluationData repositoryComponentEvaluationData =
-            repositoryComponentEvaluationDataList.componentEvalResults.get(index);
-        log.trace("  Version: {} is allowed: {}", version, !repositoryComponentEvaluationData.quarantine);
-        if (!repositoryComponentEvaluationData.quarantine) {
-          allowedVersions.put(version, versions.get(version));
-        }
-        index++;
-      }
-      componentMetadataJsonMap.put("versions", allowedVersions);
+    log.debug("Processing component metadata for {}", repoPath);
```
Comment (vladt) [inline firewall/src/main/java/com/sonatype/iq/artifactory/NpmComponentMetadataProcessor.java:61]: Only moved the try/catch up one level (to `DefaultNexusFirewallForArtifactory`)

**Review by NavyasanthiNandivelugu — APPROVED** [review]
LGTM

**Review by DanielRuaMadrid — APPROVED** [review]

File: firewall/src/main/java/com/sonatype/iq/artifactory/DefaultNexusFirewallForArtifactory.groovy [hunk 3]
```diff
@@ -235,18 +239,68 @@ class DefaultNexusFirewallForArtifactory
...
+    // Artifactory doesn't log the stacktrace for exceptions thrown from this method,
+    // so we have to use a try/catch/throw to log them here.
+    try {
+      if (initializationVerified.getCount() != 0) {
+        throw new CancelException('Invoked altRemotePathHandler before initialisation.', 500)
+      }
+
+      if (getNpmPackageNameFromRequestForPackageMetadata(repoPath) != null) {
+        repositoryManager.getArtifactoryRepositories().delete(repoPath);
+      }
```
Comment (fernau) [inline firewall/src/main/java/com/sonatype/iq/artifactory/DefaultNexusFirewallForArtifactory.groovy:287]: what would happen when pccs is disabled?

Comment (vladt) [reply firewall/src/main/java/com/sonatype/iq/artifactory/DefaultNexusFirewallForArtifactory.groovy:287]: > what would happen when pccs is disabled?

The configuration for PCCS is coming in the next story.

**Pull request merged**